delfin/drivers/hpe/hpe_3par/ssh_handler.py

```diff
@@ -71,38 +86,358 @@ def get_health_state(self):
+        if cols_size >= titles_size:
+            node_id = str_info[0]
+            cpu_info = str_info[4]
+            if obj_map.get(node_id):
+                obj_map[node_id][cpu_info] = obj_map.get(node_id).get(
+                    cpu_info, 0) + 1
+            else:
+                cpu_info_map = {}
+                cpu_info_map[cpu_info] = 1
+                obj_map[node_id] = cpu_info_map
+        return obj_map
+
+    def get_index_of_key(self, titles_list, key):
+        if titles_list:
+            for title in titles_list:
+                if key in title:
+                    return titles_list.index(title)
+        return None
+
+    @staticmethod
+    def do_exec(command_str, ssh):
+        """Execute command"""
+        try:
+            utils.check_ssh_injection(command_str)
+            if command_str is not None and ssh is not None:
+                stdin, stdout, stderr = ssh.exec_command(command_str)
+                res, err = stdout.read(), stderr.read()
+                re = res if res else err
+                result = re.decode()
+        except paramiko.AuthenticationException as ae:
+            LOG.error('doexec Authentication error:{}'.format(ae))
+            raise exception.InvalidUsernameOrPassword()
+        except Exception as e:
+            err = six.text_type(e)
+            LOG.error('doexec InvalidUsernameOrPassword error')
+            if 'timed out' in err:
+                raise exception.SSHConnectTimeout()
+            elif 'No authentication methods available' in err \
+                    or 'Authentication failed' in err:
+                raise exception.InvalidUsernameOrPassword()
+            elif 'not a valid RSA private key file' in err:
+                raise exception.InvalidPrivateKey()
+            else:
+                raise exception.SSHException(err)
+        return result
+
+    def exec_ssh_command(self, command):
+        try:
+            with self.ssh_pool.item() as ssh:
+                ssh_info = SSHHandler.do_exec(command, ssh)
+            if 'invalid command name' in ssh_info:
+                LOG.error(ssh_info)
+                raise NotImplementedError(ssh_info)
+            return ssh_info
+        except Exception as e:
+            msg = "Failed to ssh hpe 3par store %s: %s" % \
+                  (command, six.text_type(e))
+            raise exception.SSHException(msg)
+
+    def get_resources_info(self, command, analyse_type, pattern_str=None,
+                           para_map=None, ckeck_excep=True):
+        re = self.exec_ssh_command(command)
+        try:
+            if re:
+                resources_info = analyse_type(re, pattern_str,
+                                              para_map=para_map)
+        except Exception as e:
+            LOG.error("Get %s info error: %s" % (command, six.text_type(e)))
+            if ckeck_excep:
+                raise e
+        return resources_info
```

Comment:
Incase of ckeck_excep=false, in case of exception, empty lists will be returned like inventory_map
.Pls check if that is ok for all ckeck_excep=false cases.  Normally we will just allow exception to pass on if it happens

Reply:
Only the main method will throw an exception if there is an exception. The query method of other individual properties does not need to throw an exception. If throwing an exception will affect the overall return result, such as the IP address of the port, it needs another method to find out. If there is an exception in the query, the exception will be thrown, which does not affect the return of other properties

Reply:
Got the scenarios, thanks